### PR TITLE
Newloop: working filling of welds. New locks (though false-positives). 2 containers and eint (drafts but used now).

### PR DESCRIPTION
**Pull request by rfree-d**
TODO
- windows, mac problems
- unit tests for new containers and eint
- actually use/drain the weld
- actually use threads for tun read at least

### CODE REVIEW

File: src/netbuf.cpp [hunk 1]
```diff
@@ -21,20 +27,48 @@ void c_netchunk::report(std::ostream & ostr, int detail) const {
...
+	brother.m_size=0;
+	return *this;
+}
+
 c_netbuf::c_netbuf(size_t size) {
 	_dbg2("allocating");
 	m_data = new t_element[size]; // fast new - no initialization of data
+	// m_data = make_unique<t_element[]>(size);
 	m_size = size;
 	_dbg1( make_report(*this,10) );
```
Comment (robertoleksy) [inline src/netbuf.cpp]: memory leak if this will throws exception(m_data not deleted)

File: src/stdplus/arrayvector.hpp [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+		size_t m_current_size; ///< the current dynamic size, affected by push_back and pop_back etc.
+		std::array<T,TMax> m_array; ///< our actuall data, with static max size TMax
+	public:
+		T & at(size_t ix);
+		const T & at(size_t ix) const;
+		T & operator[](size_t ix);
+		const T & operator[](size_t ix) const;
+
+		void push_back(const T & val); ///< like in vector: adds one element to end, increasing size. Throws if no more space.
+		void push_back(const T && val); ///< like in vector: adds one element to end, increasing size. Throws if no more space.
```
Comment (robertoleksy) [inline src/stdplus/arrayvector.hpp]: const r-ref?

File: src/stdplus/arrayvector.hpp [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+	std::ostringstream oss;
+	oss<<"Out of range in arrayvector with TMax="<<TMax<<", m_current_size="<<m_current_size<<", requested ix="<<ix;
+	if (ix >= m_current_size) throw std::out_of_range(oss.str()); return m_array[ix];
+}
+template <typename T, size_t TMax>
+const T & arrayvector<T,TMax>::operator[](size_t ix) const {
+	std::ostringstream oss;
+	oss<<"Out of range in arrayvector with TMax="<<TMax<<", m_current_size="<<m_current_size<<", requested ix="<<ix;
+	if (ix >= m_current_size) throw std::out_of_range(oss.str()); return m_array[ix];
+}
```
Comment (robertoleksy) [inline src/stdplus/arrayvector.hpp]: Code duplication

File: src/stdplus/arrayvector.hpp [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+template <typename T, size_t TMax>
+size_t arrayvector<T,TMax>::size() const { return m_current_size; }
+
+template <typename T, size_t TMax>
+size_t arrayvector<T,TMax>::max_size() const { return TMax; }
```
Comment (robertoleksy) [inline src/stdplus/arrayvector.hpp]: constexpr

File: src/stdplus/arrayvector.hpp [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+template <typename T, size_t TMax>
+size_t arrayvector<T,TMax>::capacity() const { return TMax; }
```
Comment (robertoleksy) [inline src/stdplus/arrayvector.hpp]: constexpr

File: src/stdplus/arrayvector.hpp [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+	m_array[m_current_size] = val;
+	++m_current_size;
+}
+template <typename T, size_t TMax>
+void arrayvector<T,TMax>::push_back(const T && val) {
+	if (!(m_current_size<TMax)) {
+		std::ostringstream oss;
+		oss<<"Out of range in arrayvector with TMax="<<TMax<<", m_current_size="<<m_current_size<<", can not increase size";
+	}
+	m_array[m_current_size] = val;
```
Comment (robertoleksy) [inline src/stdplus/arrayvector.hpp]: use move

File: src/stdplus/tab.hpp [hunk 7]
```diff
@@ -106,7 +105,130 @@ typename T2::iterator copy_safe_apart(size_t n, const T1 & tab_src, T2 & tab_dst
...
+	{
+	public:
+		using iterator = T*;
+		using const_iterator = const T*;
+
+		/// of course (begin..end] must be from same container and must be a valid, linear range
+		template <typename TI>
+		tab_range(TI begin, TI end)
+				: m_begin(begin), m_end(end), m_size(end-begin) {
+			if(end<begin) throw std::invalid_argument("Incorect begin or end (end<begin).");
```
Comment (robertoleksy) [inline src/stdplus/tab.hpp]: use std::less for compare pointers

File: src/stdplus/tab.hpp [hunk 8]
```diff
@@ -106,7 +105,130 @@ typename T2::iterator copy_safe_apart(size_t n, const T1 & tab_src, T2 & tab_dst
...
+				: m_begin(obj.size()?&(obj.at(0)):nullptr), m_end(m_begin+obj.size()), m_size(obj.size()) { }
+
+		template <typename TC> // TC = type of some container. it must be linear-memory. TODO: metaprogramming check
+		tab_range(TC & obj)
+				: m_begin(obj.size()?&(obj.at(0)):nullptr), m_end(m_begin+obj.size()), m_size(obj.size()) { }
+
+		template <typename TC> // TC = type of some container. it must be linear-memory. TODO: metaprogramming check
+		tab_range(const TC & obj, size_t begin, size_t end)
+				: m_begin(&(obj.at(begin))), m_end(&(obj.at(end-1))+1), m_size(end-begin) {
+			if(end<begin) throw std::invalid_argument("Incorect begin or end (end<begin).");
```
Comment (robertoleksy) [inline src/stdplus/tab.hpp]: use std::less for compare pointers

File: src/stdplus/tab.hpp [hunk 9]
```diff
@@ -106,7 +105,130 @@ typename T2::iterator copy_safe_apart(size_t n, const T1 & tab_src, T2 & tab_dst
...
+		}
+
+		template <typename TC> // TC = type of some container. it must be linear-memory. TODO: metaprogramming check
+		tab_range(TC & obj, size_t begin, size_t end)
+				: m_begin(&(obj.at(begin))), m_end(&(obj.at(end-1))+1), m_size(end-begin) {
+			if(end<begin) throw std::invalid_argument("Incorect begin or end (end<begin).");
```
Comment (robertoleksy) [inline src/stdplus/tab.hpp]: use std::less for compare pointers

File: src/stdplus/tab.hpp [hunk 10]
```diff
@@ -106,7 +105,130 @@ typename T2::iterator copy_safe_apart(size_t n, const T1 & tab_src, T2 & tab_dst
...
+	{
+	public:
+		using iterator = T*;
+		using const_iterator = const T*;
+
+		/// of course (begin..end] must be from same container and must be a valid, linear range
+		template <typename TI>
+		tab_view(TI begin, TI end)
+				: m_begin(begin), m_end(end), m_size(end-begin) {
+			if(end<begin) throw std::invalid_argument("Incorect begin or end (end<begin).");
```
Comment (robertoleksy) [inline src/stdplus/tab.hpp]: use std::less for compare pointers

File: src/stdplus/tab.hpp [hunk 11]
```diff
@@ -106,7 +105,130 @@ typename T2::iterator copy_safe_apart(size_t n, const T1 & tab_src, T2 & tab_dst
...
+				: m_begin(obj.size()?&(obj.at(0)):nullptr), m_end(m_begin+obj.size()), m_size(obj.size()) { }
+
+		template <typename TC> // TC = type of some container. it must be linear-memory. TODO: metaprogramming check
+		tab_view(TC & obj)
+				: m_begin(obj.size()?&(obj.at(0)):nullptr), m_end(m_begin+obj.size()), m_size(obj.size()) { }
+
+		template <typename TC> // TC = type of some container. it must be linear-memory. TODO: metaprogramming check
+		tab_view(const TC & obj, size_t begin, size_t end)
+				: m_begin(&(obj.at(begin))), m_end(&(obj.at(end-1))+1), m_size(end-begin) {
+			if(end<begin) throw std::invalid_argument("Incorect begin or end (end<begin).");
```
Comment (robertoleksy) [inline src/stdplus/tab.hpp]: use std::less for compare pointers

File: src/stdplus/tab.hpp [hunk 12]
```diff
@@ -106,7 +105,130 @@ typename T2::iterator copy_safe_apart(size_t n, const T1 & tab_src, T2 & tab_dst
...
+		}
+
+		template <typename TC> // TC = type of some container. it must be linear-memory. TODO: metaprogramming check
+		tab_view(TC & obj, size_t begin, size_t end)
+				: m_begin(&(obj.at(begin))), m_end(&(obj.at(end-1))+1), m_size(end-begin) {
+			if(end<begin) throw std::invalid_argument("Incorect begin or end (end<begin).");
```
Comment (robertoleksy) [inline src/stdplus/tab.hpp]: use std::less for compare pointers

File: src/stdplus/vector_mutexed_obj.hpp [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+//	m_data.push_back(std::make_unique<TObj>(obj)); // allocate a copy, and push it
+	// @TODO@rfree bad. look at other push_back as example
+}
+*/
+
+template <typename TObj>
+void vector_mutexed_obj<TObj>::push_back(TObj && obj) {
+	UniqueLockGuardRW<MutexShared> lg_all(m_mutex_all);
+	// here TObj's move-constructor is used (stdplus::with_mutex forwards it)
+	auto new_element = std::make_unique< t_one_with_mutex >( std::move(obj) );
```
Comment (robertoleksy) [inline src/stdplus/vector_mutexed_obj.hpp]: Lock after create new_element. Critical section should be as small as possible.

File: src/stdplus/vector_mutexed_obj.hpp [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+				// unlock the one object
+			}
+		} // test all objects in loop
+		// unlocking container
+	} // access container
+	throw error_no_match_found();
+}
+
+template <typename TObj>
+size_t vector_mutexed_obj<TObj>::size() const {
```
Comment (robertoleksy) [inline src/stdplus/vector_mutexed_obj.hpp]: lock this

File: src/stdplus/with_mutex.hpp [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+};
+
+template <typename TMutex, typename TObj>
+with_mutex<TMutex,TObj>::with_mutex(const TObj & value)
+: m_obj(value)
+{ }
+
+template <typename TMutex, typename TObj>
+with_mutex<TMutex,TObj>::with_mutex(TObj && value) noexcept
+: m_obj(std::move(value))
```
Comment (robertoleksy) [inline src/stdplus/with_mutex.hpp]: use forward

File: src/stdplus/with_mutex.hpp [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+{ }
+
+template <typename TMutex, typename TObj>
+with_mutex<TMutex,TObj>::with_mutex(with_mutex<TMutex,TObj> && brother) noexcept
+// :
+// m_mutex(std::move(brother.m_mutex)),
+// m_obj(std::move(brother.m_obj))
+{
+	std::unique_lock<TMutex> lg( brother.m_mutex );
```
Comment (robertoleksy) [inline src/stdplus/with_mutex.hpp]: use unique_lock from clang wrapper

File: src/test/eint.cpp [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+template <typename T>
+static signed long long int SLL(T val) { return static_cast<signed long long int>(val); }
+
+template <typename T>
+static unsigned long long int ULL(T val) { return static_cast<unsigned long long int>(val); }
+
+TEST(eint, aggrinit_narrow_uchar) {
+	unsigned char u1=250,u2=251;
+	// unsigned char c{ u1-u2 };  expect_gt(c,0); // good: compile-error
+	EXPECT_NE(u1,u2);
```
Comment (robertoleksy) [inline src/test/eint.cpp]: very useful test

File: src/test/eint.cpp [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+TEST(eint, aggrinit_size_t) {
+	size_t a=251,b=250;
+	size_t c{ eint::eint_minus(a,b) }; // good: no-compile-error
+	// unsigned int xxx{ eint::eint_minus(a,b) }; // good: compile-error
+	EXPECT_THROW( { size_t d{ eint::eint_minus(b,a) }; EXPECT_EQ(d,b-a); } , std::range_error );
+	EXPECT_EQ(c,a-b);
+	EXPECT_EQ(ULL(c) , ULL(a)-ULL(b));
+}
+
+TEST(eint, safe_assign_from_minus) {
```
Comment (robertoleksy) [inline src/test/eint.cpp]: empty test, delete this or make this fail

File: src/ipv6.cpp [hunk 19]
```diff
@@ -0,0 +1,17 @@
...
+
+unsigned int ipv6_size_payload_from_header(stdplus::tab_view<unsigned char> tab) {
+	// https://en.wikipedia.org/wiki/IPv6_packet#Fixed_header
+	auto v_high = static_cast<unsigned int>(tab.at(4));
+	auto v_low  = static_cast<unsigned int>(tab.at(5));
+	auto val = v_low + (v_high<<8);
+	return val;
+}
+
+unsigned int ipv6_size_entireip_from_header(stdplus::tab_view<unsigned char> tab) {
```
Comment (rfree-d) [inline src/ipv6.cpp]: to platform.hpp   .. not in c++14?

File: src/mutex.hpp [hunk 20]
```diff
@@ -68,6 +71,12 @@
 #define NO_THREAD_SAFETY_ANALYSIS \
   THREAD_ANNOTATION_ATTRIBUTE__(no_thread_safety_analysis)
 
+#include <iostream>
+
+#define _mutexshield_abort(X) do { std::cerr<<"Critical error detected in mutexes: msg: " << X << " in " \
```
Comment (rfree-d) [inline src/mutex.hpp]: will move to namespace later

File: src/mutex.hpp [hunk 21]
```diff
@@ -111,41 +120,110 @@ class CAPABILITY("mutex") Mutex {
...
 };
 
 template <class t_mutex>
-class SCOPED_CAPABILITY Unique_lock {
+class SCOPED_CAPABILITY UniqueLockGuardRO {
 private:
-  std::unique_lock<t_mutex> unique_lock;
-
+  t_mutex & m_mut;
+  bool m_locked; ///< not all mutex types know if they are locked apparentyl?
```
Comment (rfree-d) [inline src/mutex.hpp]: (discussed)
we can do that, it's not data race, it's in Unique lock (not in mutex)
so we can do that by hand, and it's safe

File: src/mutex.hpp [hunk 22]
```diff
@@ -111,41 +120,110 @@ class CAPABILITY("mutex") Mutex {
...
 public:
-  Unique_lock(t_mutex &mu) ACQUIRE(mu) : unique_lock(mu) {}
+  UniqueLockGuardRO(t_mutex &mu) noexcept ACQUIRE_SHARED() : m_mut(mu), m_locked(true) {
+  	try { m_mut.lock_shared(); } catch(...) { _mutexshield_abort("LG-RO constr"); } }
+  ~UniqueLockGuardRO() noexcept RELEASE_SHARED() {
+  	try { if (m_locked) m_mut.unlock_shared(); } catch(...) { _mutexshield_abort("LG-RO destr"); } }
+
+  void lock() noexcept RELEASE_SHARED() {
```
Comment (rfree-d) [inline src/mutex.hpp]: :ok_hand:

File: src/mutex.hpp [hunk 23]
```diff
@@ -111,41 +120,110 @@ class CAPABILITY("mutex") Mutex {
...
+  	try {
+  		if ( m_locked) _mutexshield_abort("LG-RO already locked");
+	  	m_mut.unlock_shared(); m_locked=false;
+	  } catch(...) { _mutexshield_abort("LG-RO lock"); } }
+  void unlock() noexcept RELEASE_SHARED() {
+  	try {
+  		if (!m_locked) _mutexshield_abort("LG-RO already unlocked");
+	  	m_mut.unlock_shared(); m_locked=false;
```
Comment (rfree-d) [inline src/mutex.hpp]: NACK, (disc), it's unique lock (not a mutex) so it's ok here

no one can call two .unlock() on this unique lock
this classes are not thread-safe

File: src/mutex.hpp [hunk 24]
```diff
@@ -111,41 +120,110 @@ class CAPABILITY("mutex") Mutex {
...
+private:
+  t_mutex & m_mut;
+  bool m_locked; ///< not all mutex types know if they are locked apparentyl?
+public:
+  UniqueLockGuardRW(t_mutex &mu) noexcept ACQUIRE() : m_mut(mu), m_locked(true) {
+  	try { m_mut.lock(); } catch(...) { _mutexshield_abort("LG-RW constr"); } }
+  ~UniqueLockGuardRW() noexcept RELEASE() {
+  	try { if (m_locked) m_mut.unlock(); } catch(...) { _mutexshield_abort("LG-RW destr"); } }
+
+  void lock() noexcept RELEASE() {
```
Comment (rfree-d) [inline src/mutex.hpp]: :ok_hand:

File: src/mutex.hpp [hunk 25]
```diff
@@ -111,41 +120,110 @@ class CAPABILITY("mutex") Mutex {
...
+  	try {
+  		if ( m_locked) _mutexshield_abort("LG-RW already locked");
+	  	m_mut.unlock(); m_locked=false;
```
Comment (rfree-d) [inline src/mutex.hpp]: NACK, as above it's ok

File: src/netbuf.cpp (hunk 1 above)
Comment (rfree-d) [inline src/netbuf.cpp]: :ok_hand:

File: src/stdplus/arrayvector.hpp (hunk 2 above)
Comment (rfree-d) [inline src/stdplus/arrayvector.hpp]: :ok_hand:  and using now std::move inside

File: src/stdplus/arrayvector.hpp (hunk 3 above)
Comment (rfree-d) [inline src/stdplus/arrayvector.hpp]: :ok_hand: 

and removed pesimization of preparing error msg always

File: src/stdplus/arrayvector.hpp (hunk 4 above)
Comment (rfree-d) [inline src/stdplus/arrayvector.hpp]: :ok_hand:

File: src/stdplus/arrayvector.hpp (hunk 5 above)
Comment (rfree-d) [inline src/stdplus/arrayvector.hpp]: :ok_hand:

File: src/stdplus/arrayvector.hpp (hunk 6 above)
Comment (rfree-d) [inline src/stdplus/arrayvector.hpp]: used in the place with r-ref

File: src/stdplus/tab.hpp (hunk 8 above)
Comment (rfree-d) [inline src/stdplus/tab.hpp]: partially fixed; @mikurys

File: src/stdplus/tab.hpp (hunk 12 above)
Comment (rfree-d) [inline src/stdplus/tab.hpp]: partially fixed; @mikurys

File: src/stdplus/vector_mutexed_obj.hpp (hunk 13 above)
Comment (rfree-d) [inline src/stdplus/vector_mutexed_obj.hpp]: :ok_hand:

Comment (rfree-d) [inline src/stdplus/vector_mutexed_obj.hpp]: :ok_hand:

File: src/stdplus/vector_mutexed_obj.hpp (hunk 14 above)
Comment (rfree-d) [inline src/stdplus/vector_mutexed_obj.hpp]: :ok_hand:

File: src/stdplus/tab.hpp (hunk 10 above)
Comment (rfree-d) [inline src/stdplus/tab.hpp]: partially fixed; @mikurys

File: src/stdplus/with_mutex.hpp (hunk 16 above)
Comment (rfree-d) [inline src/stdplus/with_mutex.hpp]: :ok_hand:

File: src/stdplus/with_mutex.hpp (hunk 15 above)
Comment (rfree-d) [inline src/stdplus/with_mutex.hpp]: (discus.) NACK, because only one is noexecept

File: src/ipv6.cpp (hunk 19 above)
Comment (rfree-d) [inline src/ipv6.cpp]: :ok_hand:

File: src/galaxysrv_engine.hpp [hunk 26]
```diff
@@ -0,0 +1,90 @@
...
+class c_bag_fragment {
+	public:
+		weak_ptr< c_weld > m_weld;
+		size_t pos_begin, pos_end;
+};
+
+/*
+ *
+ */
+class c_bag {
```
Comment (rfree-d) [inline src/galaxysrv_engine.hpp]: :ok_hand:

File: src/galaxysrv_engine.hpp [hunk 27]
```diff
@@ -0,0 +1,90 @@
...
+
+		size_t get_free_size() const; ///< return size that is free to be written at end
+		void adjust_bufWrite(size_t new_bufWrite); ///< sets new #m_bufWrite , e.g. to correct after over-reserving size for tuntap
+};
+
+/**
+ * Queue with the Welds to be Emited.
+ * Emited from us (send from our tuntap, to next peer).
+ */
+class c_emitqueue {
```
Comment (rfree-d) [inline src/galaxysrv_engine.hpp]: :ok_hand:

File: src/galaxysrv_engine.hpp [hunk 28]
```diff
@@ -0,0 +1,90 @@
...
+/**
+ * [[optim-engine]] this marker in comments of code will mark things that could be written in other way,
+ * that could increase speed of this engine possibly. (how ever it's up for experiments to decided is it worth it)
+ * TODO
+ */
+
+/**
+ * This is both a Weld, and EI (Emit Input - buffer that forms it).
+ */
+class c_weld {
```
Comment (rfree-d) [inline src/galaxysrv_engine.hpp]: :ok_hand:

File: src/galaxysrv_engine.cpp [hunk 29]
```diff
@@ -0,0 +1,28 @@
...
+c_weld::c_weld(size_t memsize)
+:
+m_empty(true),
+// Init here things that are documented in m_empty as always valid:
+m_buf( memsize )
+{
+	Clear();
+}
+
+void c_weld::Clear() {
```
Comment (rfree-d) [inline src/galaxysrv_engine.cpp]: :ok_hand:

File: src/galaxysrv.hpp [hunk 30]
```diff
@@ -45,10 +46,11 @@
 
 #include "platform.hpp"
 
-// gal.peers.peers_add();
+#include "netbuf.hpp"
+
 
 /// @owner magaNet_user
-class c_galaxysrv : public c_galaxysrv_peers, c_galaxysrv_cables, c_galaxysrv_p2p, c_galaxysrv_e2e {
```
Comment (rfree-d) [inline src/galaxysrv.hpp]: NACK, for now decided on that design
(can re-evaluate later after this all new engine works)

File: src/galaxysrv.cpp [hunk 31]
```diff
@@ -58,77 +74,131 @@ void c_galaxysrv::main_loop() {
...
+				auto func_find =
+					[=](const c_weld & weld) {
+						auto free_size = weld.get_free_size();
+						_info("TESTING the weld "<<static_cast<const void*>(&weld)<<": free_size="<<free_size);
+						if (free_size >= cfg_max_mtu) return true;
+						return false;
+					}
+				;
+
+				auto func_use = [=](c_weld & weld) -> c_tuntap_read_result {
```
Comment (rfree-d) [inline src/galaxysrv.cpp]: to not risk modification of local vars.
will maybe change when needed

File: src/galaxysrv.cpp [hunk 32]
```diff
@@ -58,77 +74,131 @@ void c_galaxysrv::main_loop() {
...
+				struct c_tuntap_read_result {
+					c_tuntap_read_result(c_netchunk && _chunk) noexcept : chunk(std::move(_chunk)) {}
+
+					c_netchunk chunk;
+					c_haship_addr src_hip;
+					c_haship_addr dst_hip;
+				};
+
+				auto func_find =
+					[=](const c_weld & weld) {
```
Comment (rfree-d) [inline src/galaxysrv.cpp]: ditto as below

File: src/galaxysrv.cpp [hunk 33]
```diff
@@ -58,77 +74,131 @@ void c_galaxysrv::main_loop() {
...
-				c_haship_addr src_hip; // set below
-				c_haship_addr dst_hip; // set below
-				size_t read = m_tuntap.read_from_tun_separated_addresses(buf.data(), buf.size(), src_hip, dst_hip); // ***
-				c_netchunk chunk( buf.data() , read ); // actually use part of buffer
-				_info("TUN read: " << "src=" << src_hip << " " << "dst=" << dst_hip << " TUN data: " << make_report(chunk,20));
+			try {
+				_dbg2("Reading TUN...");
+
+				struct c_tuntap_read_result {
+					c_tuntap_read_result(c_netchunk && _chunk) noexcept : chunk(std::move(_chunk)) {}
```
Comment (rfree-d) [inline src/galaxysrv.cpp]: no need, tiny class, no copy-constructor.
documented, and =delete on copy

File: src/galaxysrv.cpp [hunk 34]
```diff
@@ -29,6 +30,21 @@ void c_galaxysrv::main_loop() {
 //	unique_ptr<c_cable_base_obj> cable = make_unique<c_cable_simul_obj>( world );
 //	unique_ptr<c_cable_base_addr> peer_addr = make_unique<c_cable_simul_addr>( world->generate_simul_cable() );
 
+	// ===========================================================================================================
+	_clue("Allocating in main loop");
+	size_t cfg_max_mtu = 9000;
```
Comment (rfree-d) [inline src/galaxysrv.cpp]: :ok_hand:

File: src/galaxysrv.cpp [hunk 35]
```diff
@@ -32,7 +32,7 @@ void c_galaxysrv::main_loop() {
 
 	// ===========================================================================================================
 	_clue("Allocating in main loop");
-	size_t cfg_max_mtu = 9000;
+	size_t cfg_max_mtu = 1600; // TODO read from tuntap card MTU + assert reasonable size thetn
```
Comment (robertoleksy) [inline src/galaxysrv.cpp]: 16000

File: src/galaxysrv.cpp [hunk 36]
```diff
@@ -128,7 +132,7 @@ void c_galaxysrv::main_loop() {
 					}
 					_info("TUN read: " << "src=" << tuntap_result.src_hip << " " << "dst=" << tuntap_result.dst_hip
 						<< " TUN data: " << make_report(tuntap_result.chunk,20));
-					return tuntap_result;
+					return std::move(tuntap_result);
```
Comment (robertoleksy) [inline src/galaxysrv.cpp]: why not return by value?

File: src/stdplus/tab.hpp (hunk 7 above)
Comment (mikurys) [inline src/stdplus/tab.hpp]: fixed

File: src/stdplus/tab.hpp (hunk 10 above)
Comment (mikurys) [inline src/stdplus/tab.hpp]: fixed

File: src/stdplus/tab.hpp (hunk 11 above)
Comment (mikurys) [inline src/stdplus/tab.hpp]: there is no pointers comparison

File: src/stdplus/tab.hpp (hunk 12 above)
Comment (mikurys) [inline src/stdplus/tab.hpp]: there is no pointers comparison

File: src/galaxysrv.cpp (hunk 35 above)
Comment (rfree-d) [inline src/galaxysrv.cpp]: :white:
functions to get MTU

File: src/galaxysrv.cpp (hunk 36 above)
Comment (rfree-d) [inline src/galaxysrv.cpp]: :white_check_mark: indeed

**Pull request merged**